documentation/secure_storage_rpmb.md

```diff
@@ -0,0 +1,108 @@
+ E         v          v            :         (tee_svc_storage.c)
+ R    eMMC driver   OP-TEE driver  : (tee_rpmb_fs.c, tee_fs_key_manager.c)
+ N                           ^     :            (tee_rpmb.c)
+ E                           |     :                  ^
+ L                           |     :                  |
+                             |     :                  |
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~
+                             v                        v
+                           Secure monitor / EL3 firmware
+```
+
+## The Secure Storage API
+
+This part is common with the REE-based filesystem. The interface between the
+system calls in [core/tee/tee_svc_storage.c](../core/tee/tee_svc_storage.c) and
+the RPMB filesystem is the **tee_file_operations** structure `tee_file_ops`.
+
+## The RPMB filesystem
+
+The FS implementation is mainly in [core/tee/tee_rpmb_fs.c](../core/tee/tee_rpmb_fs.c).
+
+The RPMB partition is divided in three parts:
+
+- The first 128 bytes are reserved for partition data (struct
+**rpmb_fs_partition**).
+- At offset 512 is the File Allocation Table (FAT). It is an array of
+struct **rpmb_fat_entry** elements, one per file. The FAT grows dynamically as
+files are added to the filesystem. Among other things, each entry has the start
+address for the file data, its size, and the filename.
+- Starting from the end of the RPMB partition and extending downwards is the
+file data area.
+
+Space in the partition is allocated by the general-purpose allocator functions:
+`tee_mm_alloc()` and `tee_mm_alloc2()`.
+
+## Device access
+
+The device access layer is in [core/arch/arm/tee/tee_rpmb.c](../core/arch/arm/tee/tee_rpmb.c).
+
+There is no eMMC controller driver in OP-TEE. The device operations, such as:
+programming the security key, reading the write counter value, reading or writing
+blocks of data, all have to go through the normal world. They are handled by the
+`tee-supplicant` process which further relies on the kernel's `ioctl()`
+interface to access the device. `tee-supplicant` also has an emulation mode
+which implements a virtual RPMB device for test purposes.
+
+RPMB operations are initiated on request from the FS layer. Memory buffers for
+requests and responses are allocated in shared memory using
+`thread_optee_rpc_alloc_payload()`.
+Buffers are passed to the normal world in a `TEE_RPC_RPMB_CMD` message, thanks
+to the `thread_rpc_cmd()` function. Most RPMB requests and responses use the
+data frame format defined by the JEDEC eMMC specification.
+
+HMAC authentication is implemented here also.
+
+## Encryption
+
+The FS encryption routines are in [core/tee/tee_fs_key_manager.c](../core/tee/tee_fs_key_manager.c).
+
+Block encryption is enabled by default to protect file data. It may be turned
+off with `CFG_ENC_FS=n`. The algorithm is 128-bit AES in counter (CTR) mode.
+
+- During OP-TEE initialization, a 128-bit AES Secure Storage Key (SSK) is
```

Comment:
@d3zd3z no particular reason for using CTR over CBC. I just thought it would be suitable but it seems I was wrong ;-)
So let's say we use CBC. How would you generate the IV for each RPMB block? Wikipedia more or less recommends ESSIV. Would the following work (i.e., did I understand properly)?
```
k = 128 bits of SHA256(FEK) /* hash the file encryption key */
block IV = AES_encrypt(k, 16-bit block number padded to 128 bits) /* In ECB mode? */
encrypted block data = AEC_CBC(block IV, FEK, block data)
```